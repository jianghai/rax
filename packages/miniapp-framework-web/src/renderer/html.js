/* global atagVersion */
import BASE_STYLE from '!!raw-loader!./base.css';

const POLYFILL_URL = 'https://g.alicdn.com/code/npm/@webcomponents/webcomponentsjs/2.0.4/bundles/webcomponents-sd-ce-pf.js';
const ATAG_URL = `https://g.alicdn.com/code/npm/atag/${atagVersion}/dist/atag.js`;

export function generateHTML({ viewport = 1 }) {
  return [
    '<!DOCTYPE html>',
    '<html data-iframe-root>',
    '<head>',
    '<meta charset="utf-8">',
    `<meta name="viewport" content="width=device-width, initial-scale=${viewport}, maximum-scale=${viewport}, minimum-scale=${viewport}, user-scalable=no">`,
    '<meta name="apple-mobile-web-app-capable" content="yes">',
    '<meta name="apple-mobile-web-app-status-bar-style" content="black">',
    '<meta name="format-detection" content="telephone=no">',
    `<style>${BASE_STYLE}</style>`,
    `<script src="${POLYFILL_URL}">`,
    '<\/script>',
    `<script src="${ATAG_URL}">`,
    '<\/script>',
    '</head>',
    '<body></body>',
    '</html>'].join('');
}
