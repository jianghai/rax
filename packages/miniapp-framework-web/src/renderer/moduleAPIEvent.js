/**
 * Handle with module API events.
 * @param event
 */
export default function moduleAPIEventHandler(event) {
  const { data: payload } = event;
  const { type, data } = payload;
  switch (type) {
    case 'pageScrollTo': {
      const { behavior, scrollTop } = data;
      window.scrollTo({
        top: scrollTop || 0,
        behavior: behavior || 'auto',
      });
      break;
    }
  }
}
