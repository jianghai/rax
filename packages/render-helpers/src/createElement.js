import Host from './host';
import createEmptyElement from './createEmptyElement';

/**
 * Create MiniApp element.
 * @param type {String|Object} element type.
 * @param props {Object} element props.
 * @param children {Element|Array<Element>} child elements.
 * @return {Element} element.
 */
export default function(type, props, ...children) {
  if (typeof type === 'object' && type.__esModule) {
    type = type.default;
  } else if (type === '$template') {
    if (props.is) {
      return props.is.call(props.pageInstance, props.data);
    } else {
      return createEmptyElement();
    }
  }

  // For reduce code size
  // eg: _c(View, undefined, ['xxx'])
  // vs: _c(View, ['xxx'])
  if (Array.isArray(props)) {
    children = props;
    props = undefined;
  }

  return Host.render.createElement(type, props, ...children);
}


