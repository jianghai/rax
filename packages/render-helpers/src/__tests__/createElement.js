import * as Rax from 'rax';
import renderer from 'rax-test-renderer';
import setRender from '../setRender';
import createElement from '../createElement';

setRender(Rax);
function render(el) {
  return renderer.create(el).toJSON();
}

describe('createElement', () => {
  it('can receive two args', () => {
    const el = createElement(
      'view',
      ['hello world']
    );
    expect(render(el)).toMatchSnapshot();
  });

  it('can receive three args', () => {
    const arr = [0, 1, 2];
    const el = createElement(
      'view',
      { color: 'red', arr },
      ['hello world']
    );

    // Test case for wrongly shallow copy array props, bug #872
    expect(el.props.arr).toEqual(arr);
    expect(render(el)).toMatchSnapshot();
  });

  it('can pass ES Module output object', () => {
    const el = createElement(
      { __esModule: true, default: 'view' },
      ['hello world']
    );
    expect(render(el)).toMatchSnapshot();
  });

  it('generate empty element while template is invalid', () => {
    const el = createElement(
      '$template',
      ['hello world']
    );
    expect(el).toBeNull();
  });

  it('supports template for MiniApp', () => {
    const pageInstance = Symbol('pageInstance');
    const data = Symbol('data');
    const el = createElement(
      '$template',
      {
        is(_data) {
          expect(this).toEqual(pageInstance);
          expect(_data).toEqual(data);
          return createElement('view');
        },
        pageInstance,
        data,
      }
    );
    expect(render(el)).toMatchSnapshot();
  });
});
