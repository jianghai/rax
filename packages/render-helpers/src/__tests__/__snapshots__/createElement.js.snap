// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createElement can pass ES Module output object 1`] = `
Object {
  "children": Array [
    "hello world",
  ],
  "tagName": "VIEW",
}
`;

exports[`createElement can receive three args 1`] = `
Object {
  "attributes": Object {
    "arr": Array [
      0,
      1,
      2,
    ],
    "color": "red",
  },
  "children": Array [
    "hello world",
  ],
  "tagName": "VIEW",
}
`;

exports[`createElement can receive two args 1`] = `
Object {
  "children": Array [
    "hello world",
  ],
  "tagName": "VIEW",
}
`;

exports[`createElement supports template for MiniApp 1`] = `
Object {
  "tagName": "VIEW",
}
`;
